### ISSUE TITLE
CuDNNLSTM and LSTM model weights loading, model.evaluate() issue

### ISSUE THREAD

**Issue by spate141** [keras-team/keras#9463]
- [X] Check that you are up-to-date with the master branch of Keras. You can update with:
pip install git+git://github.com/keras-team/keras.git --upgrade --no-deps

- [X] If running on TensorFlow, check that you are up-to-date with the latest version. The installation instructions can be found [here](https://www.tensorflow.org/get_started/os_setup).

I trained a Bidirectional CuDNNLSTM text classification model and `model.evaluate(x_test, y_test, batch_size=BATCH_SIZE)` give me `[('loss', 0.39137715717178684), ('acc', 0.9012292817679558)]`. Which is what I was expecting!

Now, when I load the same model weights in a new Bidirectional LSTM model (to run the model on CPU) by modifying `CuDNNLSTM > LSTM` in my model architecture, I'm getting `[('loss', 8.747908523430075), ('acc', 0.006823506011315417)]` on same test data.

Any suggestion on this behavior? Do I have to update anything else when loading CuDNNLSTM model weights in LSTM model? I did the same thing for my other model and it worked without any issues. 

Also, parameters numbers seems to be different in CuDNNLSTM and LSTM layer, even though everything is same! 

**CuDNNLSTM:** 
![image](https://user-images.githubusercontent.com/10580847/36563351-796724ec-17df-11e8-9517-be4ddd15dff0.png)

**LSTM:**
![image](https://user-images.githubusercontent.com/10580847/36563359-7d192338-17df-11e8-85fd-5c1fdf81904a.png)

**Comment by spate141** [comment]
CuDNNLSTM model confusion matrix:
![image](https://user-images.githubusercontent.com/10580847/36563607-4f0c4dca-17e0-11e8-84d5-fe46302c4355.png)

After loading CuDNNLSTM model weights in LSTM model confusion matrix:
![image](https://user-images.githubusercontent.com/10580847/36563636-6919ad8e-17e0-11e8-84ae-e9055006d1cc.png)

**Comment by bzamecnik** [comment]
Oh!

**Comment by myutwo150** [comment]
I'm not able to reproduce this error. Which Keras version do you use? 

Does the following code passes on your machine?

```python
model = Sequential([
    Embedding(100, 10, input_shape=(20,)),
    Bidirectional(CuDNNLSTM(10)),
    BatchNormalization(),
    Dropout(0.2),
    Dense(10)
])
model.save_weights('1.h5')
model.compile(loss='mse', optimizer='adam')

m = Sequential([
    Embedding(100, 10, input_shape=(20,)),
    Bidirectional(LSTM(10, recurrent_activation='sigmoid')),
    BatchNormalization(),
    Dropout(0.2),
    Dense(10)
])
m.load_weights('1.h5')
m.compile(loss='mse', optimizer='adam')

X, y = np.random.randint(100, size=(100, 20)), np.random.rand(100, 10)
assert model.evaluate(X, y) == m.evaluate(X, y)
```

**Comment by spate141** [comment]
I'm using `Keras: 2.1.4` and above code pass on p2.xlarge instance that i'm using for training.

![image](https://user-images.githubusercontent.com/10580847/36577526-73071ae2-181d-11e8-94ce-338d4c3a3264.png)

However, as I can't compile `CuDNNLSTM` based model on CPU, I can't check the `assert` condition. But, I was able to load the weights from `CuDNNLSTM` to `LSTM` model on CPU based instance successfully, as I have mentioned the same on my previous comment showing confusion matrix after loading weights from `CuDNNLSTM to LSTM` without any error!

![image](https://user-images.githubusercontent.com/10580847/36577762-c043bb02-181e-11e8-845f-e2e35292f451.png)

**Comment by spate141** [comment]
#### UPDATE 🎈

After using `sigmoid` instead of default `hard_sigmoid` recurrent_activation function in LSTM layer, I was able to get acceptable results! 🕺🏻

I'm surprised that I missed this as @bzamecnik also mentioned about the same in #8860! Also thanks to @myutwo150 for pointing out the obvious with the example! 

![image](https://user-images.githubusercontent.com/10580847/36578181-2a0aef5e-1821-11e8-9aa1-1db52569ec49.png)

**Comment by danielcmm** [comment]
I still can't get it to work... loading CuDNNLSTM weigths into a LSTM model.
Although i can get it to load and run, the accuracy drops from **0.76 with CUDNN** to **0.13 without**.



Here is my code:

```
weights_file = "w.h5"
units1 = 512
units2 = 256
tamanho_vocab = 50000
tamanho_vetor_w2v = 350
```

```
cudnn_model = Sequential([
    Embedding(input_dim=tamanho_vocab, output_dim=tamanho_vetor_w2v, weights=[pretrained_weights]),
    CuDNNLSTM(units1, return_sequences=True),
    Dropout(0.1),
    CuDNNLSTM(units=units2),
    Dropout(0.1),
    Dense(tamanho_vocab, activation='softmax')  
])
cudnn_model.compile(loss='sparse_categorical_crossentropy', optimizer='adam', metrics=["accuracy"])

# Train, fit, etc...
cudnn_model.save_weights(weigths_file)
```

And on another machine with only CPU...

```
model = Sequential([
    Embedding(input_dim=tamanho_vocab, output_dim=tamanho_vetor_w2v, weights=[pretrained_weights]),
    LSTM(units1, return_sequences=True ,recurrent_activation="sigmoid"),
    Dropout(0.1),
    LSTM(units=units2,recurrent_activation="sigmoid"),
    Dropout(0.1),
    Dense(tamanho_vocab, activation='softmax')  
])
```

```
model.load_weights(weigths_file)
model.compile(loss='sparse_categorical_crossentropy', optimizer='adam', metrics=["accuracy"])
```

Im using tensorflow 1.10.1 and have already tried with recurrent_activation = "hard_sigmoid" and it just gets worse.

Any ideas?

thanks!

**Comment by illi4** [comment]
Have the same issue. If anyone could suggest anything, it would be great. 

> I still can't get it to work... loading CuDNNLSTM weigths into a LSTM model.
> Although i can get it to load and run, the accuracy drops from **0.76 with CUDNN** to **0.13 without**.

**Comment by yogurfrul** [comment]
do u have set the CuDNNLSTM time-major?
normal input:
[batch_size, time_len, input_size] 
---------
cudnn_lstm input:
 [time_len, batch_size, input_size]

**Comment by wangye8899** [comment]
CuDNNLSTM is the timesteps-major,you should reshape your inputs [timesteps,batch,input_size] , and finally reshape outputs to [batch,timesteps,input_size]

**Comment by mcourteaux** [comment]
> CuDNNLSTM is the timesteps-major,you should reshape your inputs [timesteps,batch,input_size] , and finally reshape outputs to [batch,timesteps,input_size]

The original tf.contrib module indeed had [time, batch, feature], but Keras transposes this for you under the hood. Check the source here: https://github.com/keras-team/keras/blob/master/keras/layers/cudnn_recurrent.py#L484
